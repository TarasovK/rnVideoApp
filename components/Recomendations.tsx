import React from 'react'
import { StyleSheet, FlatList, View, Dimensions, ViewToken } from 'react-native'
import { VideoPlayer } from './VideoPlayer'
import { SideBar } from './SideBar'
import VideoContentApi from '../api/VideoContentApi'

const windowWidth = Dimensions.get('window').width;
const windowHeight = Dimensions.get('window').height;
const viewabilityConfig = {
    itemVisiblePercentThreshold: 50,
  };

export interface Props {
    onAuthorPress?: (() => void)
    onCommentsPress?: (() => void)
    onLikePress?: (() => void)
    onSharePress?: (() => void)
}

export class Recomendations extends React.Component<Props> {
    cellRefs: Array<{playerRef: VideoPlayer | null}>
    
    constructor(props: Props) {
        super(props)
        this.cellRefs = new Array
    }   

    renderItem = ({ item }: {item: {id: string, uri: string, poster: string}}) => {
        return(
            <View style={styles.listItem}>            
                <VideoPlayer 
                    ref={ref=> this.cellRefs[parseInt(item.id)] = {playerRef: ref}}
                    uri={item.uri}/>
                <View style={styles.sidebarView}>
                    <SideBar onCommentsPress={this.props.onCommentsPress}/>
                </View>                   
            </View>
        )
    }

    onViewableItemsChanged = (info: {viewableItems: ViewToken[], changed: ViewToken[]}) => {
        info.changed.forEach(element => {
            var cell = this.cellRefs[parseInt(element.key)];
            if(cell) {
                element.isViewable ? cell.playerRef!.play() : cell.playerRef!.pause()
            }
        });
    }
    
    render() {
        return(
            <View style={styles.root}>
                <FlatList 
                    data={VideoContentApi.getVideoContent()}
                    renderItem={this.renderItem}
                    showsVerticalScrollIndicator={false}  
                    onViewableItemsChanged={this.onViewableItemsChanged}  
                    viewabilityConfig={viewabilityConfig}               
                />                
            </View>
        )
    }
}

var styles = StyleSheet.create({
    root: {
        flex: 1,
        justifyContent: "center",
        alignItems: "center",
        borderColor: '#0FF',
        borderWidth: 0,
        zIndex: 0
    },
    listItem: {
        flex: 1,
        width: windowWidth,
        height: windowHeight
    },   
    sidebarView: {
        width: 80,
        height: 320,
        position: "absolute",
        right: 0,
        top: windowHeight / 2 - 160
    },
})