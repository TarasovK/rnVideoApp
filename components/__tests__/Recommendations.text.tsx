import 'react-native';
import React from 'react';
import { Recomendations }  from '../Recomendations';

import renderer from 'react-test-renderer';

jest.mock('react-native-video', () => 'Video')

it('renders correctly', () => {
  const tree = renderer.create(<Recomendations />);
  expect(tree).toMatchSnapshot()
});