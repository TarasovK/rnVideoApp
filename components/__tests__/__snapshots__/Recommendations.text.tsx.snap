// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "borderColor": "#0FF",
      "borderWidth": 0,
      "flex": 1,
      "justifyContent": "center",
      "zIndex": 0,
    }
  }
>
  <RCTScrollView
    data={
      Array [
        Object {
          "id": "0",
          "poster": "https://commondatastorage.googleapis.com/gtv-videos-bucket/sample/images/ForBiggerBlazes.jpg",
          "uri": "https://commondatastorage.googleapis.com/gtv-videos-bucket/sample/ForBiggerBlazes.mp4",
        },
        Object {
          "id": "1",
          "poster": "https://commondatastorage.googleapis.com/gtv-videos-bucket/sample/images/BigBuckBunny.jpg",
          "uri": "https://commondatastorage.googleapis.com/gtv-videos-bucket/sample/BigBuckBunny.mp4",
        },
        Object {
          "id": "2",
          "poster": "https://commondatastorage.googleapis.com/gtv-videos-bucket/sample/images/ForBiggerMeltdowns.jpg",
          "uri": "https://commondatastorage.googleapis.com/gtv-videos-bucket/sample/ForBiggerMeltdowns.mp4",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    onViewableItemsChanged={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    showsVerticalScrollIndicator={false}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfig={
      Object {
        "itemVisiblePercentThreshold": 50,
      }
    }
    viewabilityConfigCallbackPairs={
      Array [
        Object {
          "onViewableItemsChanged": [Function],
          "viewabilityConfig": Object {
            "itemVisiblePercentThreshold": 50,
          },
        },
      ]
    }
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "flex": 1,
              "height": 1334,
              "width": 750,
            }
          }
        >
          <Video
            paused={true}
            repeat={true}
            resizeMode="contain"
            source={
              Object {
                "uri": "https://commondatastorage.googleapis.com/gtv-videos-bucket/sample/ForBiggerBlazes.mp4",
              }
            }
            style={
              Object {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              }
            }
          />
          <View
            style={
              Object {
                "height": 320,
                "position": "absolute",
                "right": 0,
                "top": 507,
                "width": 80,
              }
            }
          >
            <View
              style={
                Object {
                  "flex": 1,
                  "flexDirection": "column",
                }
              }
            >
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "opacity": 1,
                  }
                }
              >
                <Image
                  source={
                    Object {
                      "testUri": "../../../components/icons/author.png",
                    }
                  }
                  style={
                    Object {
                      "flex": 1,
                      "resizeMode": "contain",
                    }
                  }
                />
                <Text
                  style={
                    Object {
                      "color": "#FFF",
                      "flex": 1,
                      "fontSize": 12,
                      "textAlign": "center",
                    }
                  }
                >
                  Author
                </Text>
              </View>
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "opacity": 1,
                  }
                }
              >
                <Image
                  source={
                    Object {
                      "testUri": "../../../components/icons/like.png",
                    }
                  }
                  style={
                    Object {
                      "flex": 1,
                      "resizeMode": "contain",
                    }
                  }
                />
                <Text
                  style={
                    Object {
                      "color": "#FFF",
                      "flex": 1,
                      "fontSize": 12,
                      "textAlign": "center",
                    }
                  }
                >
                  1.8m
                </Text>
              </View>
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "opacity": 1,
                  }
                }
              >
                <Image
                  source={
                    Object {
                      "testUri": "../../../components/icons/comments.png",
                    }
                  }
                  style={
                    Object {
                      "flex": 1,
                      "resizeMode": "contain",
                    }
                  }
                />
                <Text
                  style={
                    Object {
                      "color": "#FFF",
                      "flex": 1,
                      "fontSize": 12,
                      "textAlign": "center",
                    }
                  }
                >
                  Comments
                </Text>
              </View>
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "opacity": 1,
                  }
                }
              >
                <Image
                  source={
                    Object {
                      "testUri": "../../../components/icons/share.png",
                    }
                  }
                  style={
                    Object {
                      "flex": 1,
                      "resizeMode": "contain",
                    }
                  }
                />
                <Text
                  style={
                    Object {
                      "color": "#FFF",
                      "flex": 1,
                      "fontSize": 12,
                      "textAlign": "center",
                    }
                  }
                >
                  Share
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "flex": 1,
              "height": 1334,
              "width": 750,
            }
          }
        >
          <Video
            paused={true}
            repeat={true}
            resizeMode="contain"
            source={
              Object {
                "uri": "https://commondatastorage.googleapis.com/gtv-videos-bucket/sample/BigBuckBunny.mp4",
              }
            }
            style={
              Object {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              }
            }
          />
          <View
            style={
              Object {
                "height": 320,
                "position": "absolute",
                "right": 0,
                "top": 507,
                "width": 80,
              }
            }
          >
            <View
              style={
                Object {
                  "flex": 1,
                  "flexDirection": "column",
                }
              }
            >
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "opacity": 1,
                  }
                }
              >
                <Image
                  source={
                    Object {
                      "testUri": "../../../components/icons/author.png",
                    }
                  }
                  style={
                    Object {
                      "flex": 1,
                      "resizeMode": "contain",
                    }
                  }
                />
                <Text
                  style={
                    Object {
                      "color": "#FFF",
                      "flex": 1,
                      "fontSize": 12,
                      "textAlign": "center",
                    }
                  }
                >
                  Author
                </Text>
              </View>
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "opacity": 1,
                  }
                }
              >
                <Image
                  source={
                    Object {
                      "testUri": "../../../components/icons/like.png",
                    }
                  }
                  style={
                    Object {
                      "flex": 1,
                      "resizeMode": "contain",
                    }
                  }
                />
                <Text
                  style={
                    Object {
                      "color": "#FFF",
                      "flex": 1,
                      "fontSize": 12,
                      "textAlign": "center",
                    }
                  }
                >
                  1.8m
                </Text>
              </View>
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "opacity": 1,
                  }
                }
              >
                <Image
                  source={
                    Object {
                      "testUri": "../../../components/icons/comments.png",
                    }
                  }
                  style={
                    Object {
                      "flex": 1,
                      "resizeMode": "contain",
                    }
                  }
                />
                <Text
                  style={
                    Object {
                      "color": "#FFF",
                      "flex": 1,
                      "fontSize": 12,
                      "textAlign": "center",
                    }
                  }
                >
                  Comments
                </Text>
              </View>
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "opacity": 1,
                  }
                }
              >
                <Image
                  source={
                    Object {
                      "testUri": "../../../components/icons/share.png",
                    }
                  }
                  style={
                    Object {
                      "flex": 1,
                      "resizeMode": "contain",
                    }
                  }
                />
                <Text
                  style={
                    Object {
                      "color": "#FFF",
                      "flex": 1,
                      "fontSize": 12,
                      "textAlign": "center",
                    }
                  }
                >
                  Share
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "flex": 1,
              "height": 1334,
              "width": 750,
            }
          }
        >
          <Video
            paused={true}
            repeat={true}
            resizeMode="contain"
            source={
              Object {
                "uri": "https://commondatastorage.googleapis.com/gtv-videos-bucket/sample/ForBiggerMeltdowns.mp4",
              }
            }
            style={
              Object {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              }
            }
          />
          <View
            style={
              Object {
                "height": 320,
                "position": "absolute",
                "right": 0,
                "top": 507,
                "width": 80,
              }
            }
          >
            <View
              style={
                Object {
                  "flex": 1,
                  "flexDirection": "column",
                }
              }
            >
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "opacity": 1,
                  }
                }
              >
                <Image
                  source={
                    Object {
                      "testUri": "../../../components/icons/author.png",
                    }
                  }
                  style={
                    Object {
                      "flex": 1,
                      "resizeMode": "contain",
                    }
                  }
                />
                <Text
                  style={
                    Object {
                      "color": "#FFF",
                      "flex": 1,
                      "fontSize": 12,
                      "textAlign": "center",
                    }
                  }
                >
                  Author
                </Text>
              </View>
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "opacity": 1,
                  }
                }
              >
                <Image
                  source={
                    Object {
                      "testUri": "../../../components/icons/like.png",
                    }
                  }
                  style={
                    Object {
                      "flex": 1,
                      "resizeMode": "contain",
                    }
                  }
                />
                <Text
                  style={
                    Object {
                      "color": "#FFF",
                      "flex": 1,
                      "fontSize": 12,
                      "textAlign": "center",
                    }
                  }
                >
                  1.8m
                </Text>
              </View>
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "opacity": 1,
                  }
                }
              >
                <Image
                  source={
                    Object {
                      "testUri": "../../../components/icons/comments.png",
                    }
                  }
                  style={
                    Object {
                      "flex": 1,
                      "resizeMode": "contain",
                    }
                  }
                />
                <Text
                  style={
                    Object {
                      "color": "#FFF",
                      "flex": 1,
                      "fontSize": 12,
                      "textAlign": "center",
                    }
                  }
                >
                  Comments
                </Text>
              </View>
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "opacity": 1,
                  }
                }
              >
                <Image
                  source={
                    Object {
                      "testUri": "../../../components/icons/share.png",
                    }
                  }
                  style={
                    Object {
                      "flex": 1,
                      "resizeMode": "contain",
                    }
                  }
                />
                <Text
                  style={
                    Object {
                      "color": "#FFF",
                      "flex": 1,
                      "fontSize": 12,
                      "textAlign": "center",
                    }
                  }
                >
                  Share
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
